text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,278 @@
+In that case, the `keep-stdlib-dependencies` will continue to keep the builds of users working as is, but since the prohibition of explicit stdlib deps on crates.io will disappear, `custom-implicit-stdlib-dependencies` can be removed.
+
+Once the remaining aspects of this are implemente---the binary mock source and lockfile filtering of stdlib deps, and rustbuild uses the new system---the `keep-stdlib-dependencies` key can also be removed: the pruning key can be removed because the mock sysroot registry will offer a wholly safer way to use pre-built binaries.
+If, at that point, there is any official policy on how many prior Rust versions official crates need to work with, the removal can be delayed.
+This will allow official crates to use explicit deps without breaking the old versions of Cargo those supported old Rust releases were distributed with.
+
+
+# Drawbacks
+
+ - The sysroot binary mock source is a complicated special case whose implementation will probably span many parts of Cargo.
+   In the near future, it is unlikely to be generalized into something more elegant.
+
+ - Even with this RFC and a nightly compiler, a single `cargo build` is incapable of building the entire standard library due to external dependencies.
+   But I believe we will eventually reach that goal, and furthermore this RFC will help us reach it.
+
+ - Compilers could provide crates in their sysroot that don't match the Rust specification, and Cargo would be none the wiser.
+   (Technically, this problem already exists with falling back on the sysroot binaries, but users will probably expect better when they can specify standard library dependencies explicitly.)
+   Since the *interface* of the stdlib is specified, it would be neat if we could put a big crate type/interface on crates.io, which compiler implementations would need to match.
+   [That is, the interface of the stable crates.
+   Unstable crates behind the std facade are a compiler-specific implementation detail, and thus it would be counter-productive, even to likewise constrain their interfaces.]
+
+ - The name "compiler source" is unfortunate because it sounds like the source of the compiler itself.
+   But perhaps the best solution is to rename Cargo's "source" abstraction (and the traits that go with it).
+
+
+# Alternatives
+
+ - Previous versions of this RFC were simpler but more brittle.
+   Please refer to the git history to see them.
+
+ - Should the "stdlib" virtual source instead be called "sysroot" (e.g. `core =  { sysroot = true, .. }`)?
+   This emphasizes how those dependencies are resolved as opposed to what they are for.
+
+ - If a way to specify the language version like #1709 or #1707 is added, the version of stdlib dependencies could be pulled from that.
+
+ - This has been deemed sufficiently complicated to warrant the introduction of unstable features to Cargo.
```

Comment:
"This has been deemed sufficiently complicated..."  -- I think that sentence actually belongs in the "Drawbacks" section.